Skipchain propagation optimization

Gilthoniel opened the pull request:
This changes the way the skipchain is propagating new block and
forward links so that it sends minimal information and let conodes
ask for missing data.

Fixes #1395 #1213 #1201

jeffallen commented on skipchain/skipchain.go:
```diff
@@ -1360,22 +1349,76 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
+	// Because of the sync, the forward link could already be added
+	// but we need to catch the case where the height is much above
+	// the current one which means we missed previous propagation
+	if len(sb.ForwardLink) <= pfl.Height {
```
Did we have a test case for this before? Did you have to make one?

Gilthoniel replied:
I had errors after I changed the propagation test which are actually not errors. As the comment says, if you don't have the previous block and then try to sync, you might end up with the block already updated by another member of the cothority. In my case the leader was sending the block with the forward link already updated.

jeffallen commented on skipchain/skipchain.go:
```diff
@@ -1360,22 +1349,76 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
+// PropagateForwardLinkHandler will update the latest block with
+// the new forward link and the new block when given
+func (s *Service) propagateForwardLinkHandler(msg network.Message) {
+	pfl, ok := msg.(*PropagateForwardLink)
```
We need to think about security here... pfl is essentially untrusted input from the network at this point. Anyone who manages to send the right message into onet can get us to execute this code with their input. What bad stuff can they do? (This was already true before, but I'm wondering if somehow this change opens us up to trying SyncChains we shouldn't be trying?)

Gilthoniel replied:
The roster could actually be taken from the previous blocks. That would fix the three points you pointed out.

jeffallen commented on skipchain/skipchain.go:
```diff
@@ -1360,22 +1349,76 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
+	sb := s.db.GetByID(pfl.ForwardLink.From)
+	if sb == nil {
+		// Happen when the roster has changed
+		err := s.SyncChain(pfl.Roster, pfl.ForwardLink.From)
```
pfl.Roster is controlled by the attacker here, that doesn't seem right; they could ask us to try to go talk to whitehouse.gov!

ineiti replied:
I like the reference:
- what is the worst that could happen in this code?
- DDoS attack on whitehouse.gov!
- oh dang - the world is going to end!

Besides the funny reference you're of course completely right.

jeffallen replied:
Honestly, right now a DOS on whitehouse.gov's access to twitter.com might be an improvement.

jeffallen commented on skipchain/skipchain.go:
```diff
@@ -1360,22 +1349,76 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
+	}
+
+	// Update the forward link list of the block
+	s.db.Store(sb)
```
I'm not clear where this block was validated before we accept it. Does that happen in some other layer of the code? Or is forgotten here?

ineiti replied:
As written above, the block should not even be propagated here. Only taken from the cache (populated by `bftForwardLink`) and then stored if the forward link works out.

Gilthoniel replied:
the skipchain service Store function updates existing blocks with the new forward links after verifying them. If the block as been tempered, the hash is different and it won't be found in the db, then it won't be stored because it can be linked.

That could actually be improved by checking if the id is returned. If not the data is corrupted.

jeffallen commented on skipchain/skipchain_test.go:
```diff
@@ -609,12 +609,44 @@ func TestService_Propagation(t *testing.T) {
 	service := genService.(*Service)
 
-	sbRoot, err := makeGenesisRosterArgs(service, ro, nil, VerificationNone,
+	roster := onet.NewRoster(ro.List[:5])
```
it would be easier to understand this test if these were renamed as: ro -> fullRoster, roster -> firstFiveRoster

ineiti commented on skipchain/skipchain.go:
```diff
@@ -975,27 +958,16 @@ func (s *Service) forwardLinkLevel0(src, dst *SkipBlock) error {
+	// could have changed, meaning one node cannot know the block
+	return s.startForwardLinkPropagation(ro, &PropagateForwardLink{
+		ForwardLink: fwd,
+		Newest:      dst,
```
Needed: For optimizing it correctly, you have to store the block in the `bftForwardLinkLevel0`, so this will not send it again.

ineiti commented on skipchain/skipchain.go:
```diff
@@ -975,27 +958,16 @@ func (s *Service) forwardLinkLevel0(src, dst *SkipBlock) error {
+	return s.startForwardLinkPropagation(ro, &PropagateForwardLink{
+		ForwardLink: fwd,
+		Newest:      dst,
+		Roster:      roster,
```
Question: I still don't see why you send the roster here.

ineiti commented on skipchain/skipchain.go:
```diff
@@ -1360,22 +1349,76 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
-		if !s.blockIsFriendly(sb) {
-			log.Lvlf2("%s: block is not friendly: %x", s.ServerIdentity(), sb.Hash)
+
+	s.db.Store(pg.Genesis)
```
Please use `blockIsFriendly` here to make sure we should accept the block. It is a very simple authorization mechanism that allows to set up following of chains and nodes.

ineiti commented on skipchain/skipchain.go:
```diff
@@ -1360,22 +1349,76 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
+
+	// Get the block to update the list of FLs
+	sb := s.db.GetByID(pfl.ForwardLink.From)
+	if sb == nil {
```
As far as I understand, if `sb == nil`, then we can return here. What usually happens is this (for level >0 link):
1. leader node asks earlier leader with `ForwardSignature` to create a forward signature
2. earlier leader creates forward signature using `forwardLink`
3. all earlier nodes verify the forward-link and _TODO_: store the block
4. the earlier leader sends `propagateForwardLinkHandler` with the forwardlink

So if a node receives a forward-link
- pointing to/from a block it doesn't has, it can ignore it
- it cannot verify, it can ignore it

If you're in doubt, write me on slack, and we'll whiteboard it.

kc1212 commented on skipchain/struct.go:
```diff
@@ -485,6 +485,58 @@ func (sb *SkipBlock) updateHash() SkipBlockID {
+				return errors.New("Missing forward links")
+			}
+
+			fl := sb.ForwardLink[len(sb.ForwardLink)-1]
```
are you not checking all the forward links?

Gilthoniel replied:
Only the highest one because the block for the lower ones are not propagated

Gilthoniel commented:
Addressing the comments required some major changes in order to cache the newest block and store it after it has been verified. i guess you should go over it once again.

ineiti commented on skipchain/skipchain.go:
```diff
@@ -1360,23 +1369,117 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
-			log.Lvlf2("%s: block is not friendly: %x", s.ServerIdentity(), sb.Hash)
+
+	if !s.blockIsFriendly(pg.Genesis) {
+		log.Error("Block is not friendly")
```
```suggestion
		log.Error("Conode doesn't want to follow that skipchain")
```

ineiti commented on skipchain/skipchain.go:
```diff
@@ -1360,23 +1369,117 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
+
+	// Add the block if available
+	if sb, ok := s.blockBuffer.get(sb.SkipChainID(), pfl.ForwardLink.To); ok {
+		if !s.blockIsFriendly(sb) {
```
Is that possible? We have the genesis block, so I think `blockIsFriendly` will return true anyway.

ineiti commented on skipchain/skipchain.go:
```diff
@@ -1360,23 +1369,117 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
+}
+
+// TODO: change messaging package to allow to return an error
+// it makes the function easier to test
```
Problem here is that the messages are asynchronous - we don't have a way to return an error to the sender.

ineiti replied:
OK - sounds good. Opened dedis/onet#501

ineiti commented on skipchain/struct.go:
```diff
@@ -1027,3 +1112,57 @@ func (db *SkipBlockDB) getAllSkipchains() (map[string]*SkipBlock, error) {
+
+func newSkipBlockBuffer() *skipBlockBuffer {
+	return &skipBlockBuffer{
+		buffer: make(map[[32]byte]*SkipBlock),
```
We're using `string` as key in other places - @jeffallen can surely tell us which one is better: using `string` as key or array of bytes?

ineiti commented on skipchain/struct.go:
```diff
@@ -1027,3 +1112,57 @@ func (db *SkipBlockDB) getAllSkipchains() (map[string]*SkipBlock, error) {
+
+// get returns the block if the skipchain ID hits with the
+// correct block ID
+func (sbb *skipBlockBuffer) get(sid SkipBlockID, id SkipBlockID) (*SkipBlock, bool) {
```
Why do you return a `bool` to indicate if the block is available or not? Isn't a `== nil` the same?
